### PR TITLE
add gpt sft

### PR DESCRIPTION
**Pull request by JustinPan-goog**
| Source Files | Reference Files | Reference Functions | Usage |
| -------- | ------- | ------- | ------- |
| `input_pipeline_gpu.py` | `nemo_aligner.utils` | `parallel_state` | Wrapper around mcore parallel state to handle cases of resharding |
| | | | |
| `model_utils_gpu.py` | `nemo_aligner.utils.train_script_utils` | `init_distributed` | Initialize the distributed process group with help from the PTL Trrainer |
| `model_utils_gpu.py` | `nemo_aligner.utils.train_script_utils` | `init_peft` | initialize peft weights |
| `model_utils_gpu.py` | `nemo_aligner.utils.utils` | `load_from_nemo` | load a model using nemo checkpoint |
| `model_utils_gpu.py` | `nemo_aligner.models.nlp.gpt.gpt_sft_model` | `GPTSFTModel` | SFT Model |
| | | | |
| `trainer_utils_gpu.py` | `nemo_aligner.metrics` | `InferenceMetricsHandler` | Wrapper around metrics objects that will call update/compute/reset on all registered metrics |
| `trainer_utils_gpu.py` | `nemo_aligner.utils.distributed` | `SyncTimer` | Wrapper around NamedTimer to sync across DP ranks for more precise timing |
| `trainer_utils_gpu.py` | `nemo_aligner.utils.train_utils` | `clip_gradients` | Gradient clipping |
| `trainer_utils_gpu.py` | `nemo_aligner.utils.trainer_utils` | `check_progress` | Trainer utilities |
| `trainer_utils_gpu.py` | `nemo_aligner.utils.trainer_utils` | `compute_limit_batches` | Trainer utilities |
| `trainer_utils_gpu.py` | `nemo_aligner.utils.trainer_utils` | `compute_num_steps_per_epoch` | Trainer utilities |

### CODE REVIEW

File: clm/gpt_sft_dataset.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+from typing import Mapping, Optional
```
Comment (ZhiyuLi-goog) [inline clm/gpt_sft_dataset.py:1]: In addition to `GPTSFTDataset`, we might probably need to support [`GPTDataset`](https://github.com/NVIDIA/NeMo/blob/7a3dd6bfe17c0ff16abf04b55c37a5e56f564227/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py#L44) as well.
`GPTDataset` might be a pure document text without mutli-run chat template.


Some background
*  [c4 dataset](https://github.com/mlcommons/training/blob/master/large_language_model/megatron-lm/pretrain_gpt.py#L117-L135)
* [nvidia's previous submission](https://github.com/mlcommons/training_results_v3.1/blob/main/NVIDIA/benchmarks/gpt3/implementations/pytorch/custom_callbacks.py#L33)
* [Google's previous submission](https://github.com/AI-Hypercomputer/maxtext/blob/77f079f845084ba853453dc0755deb0daf312e26/MaxText/input_pipeline/_tfds_data_processing_c4_mlperf.py#L230)
* [sample of c4 dataset](https://huggingface.co/datasets/allenai/c4)

Comment (ZhiyuLi-goog) [reply clm/gpt_sft_dataset.py:1]: Might not necessary in this PR. Feel free to add a new PR.

File: clm/model_utils_gpu.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
+import os
+from omegaconf.omegaconf import OmegaConf, open_dict
+import torch
+
+from nemo.utils import logging
+from nemo.collections.nlp.data.language_modeling.megatron.gpt_sft_chat_dataset import get_prompt_template_example
+
+from nemo_aligner.utils.train_script_utils import (
+    init_distributed,
+    init_peft,
```
Comment (ZhiyuLi-goog) [inline clm/model_utils_gpu.py:10]: `init_peft`(parameter effecient finetuning, say [LoRA](https://arxiv.org/abs/2106.09685) as an example) should be 100% out of the scope, can we simplify the pipeline by removing all the relevant features?

File: clm/model_utils_gpu.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    model, updated_cfg = load_from_nemo(
+        GPTSFTModel,
+        config,
+        trainer,
+        strict=True,
+        modify_config_fn=_modify_config,
+        restore_path=config.model.restore_from_path,
+        return_updated_cfg=True,
+    )
+    init_peft(model, updated_cfg)
```
Comment (ZhiyuLi-goog) [inline clm/model_utils_gpu.py:54]: We don't need `peft` in the scope.

File: clm/model_utils_gpu.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            gpt_cfg.pipeline_model_parallel_size = cfg.model.get("pipeline_model_parallel_size", 1)
+        gpt_cfg.pipeline_model_parallel_split_rank = cfg.model.get("pipeline_model_parallel_split_rank", 0)
+
+        if cfg.model.data.get("chat", False):
+            # chat model, overwrite the prompt template
+            prompt_template = get_prompt_template_example(cfg.model.data.chat_prompt_tokens)
+            gpt_cfg.data.train_ds.prompt_template = prompt_template
+            gpt_cfg.data.validation_ds.prompt_template = prompt_template
+
+        sft_cls = GPTSFTModel
```
Comment (ZhiyuLi-goog) [inline clm/model_utils_gpu.py:113]: At first glance, looks like `GPTSFTModel` is a class in nemo aligner which covers `NLPAdapterModelMixin` specifically [here](https://github.com/NVIDIA/NeMo-Aligner/blob/d8ef9fb05dc9a30780d5f0092b918476fecd83be/nemo_aligner/models/nlp/gpt/gpt_sft_model.py#L49)

Given no need to cover PEFT adapter, we can directly use `MegatronGPTModel` instead if possible.
* [training `MegatronGPTModel`](https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_pretraining.py)
* [mlperf example `MegatronGPTModel`](https://github.com/mlcommons/training_results_v3.1/blob/main/NVIDIA/benchmarks/gpt3/implementations/pytorch/custom_callbacks.py#L523)

If we we need some api in `SupervisedInterface` feel free to leave as it is.

File: clm/input_pipeline_gpu.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
+import torch
+from gpt_sft_dataset import GPTSFTDatasetHF
+
+from nemo.utils import logging
+from nemo_aligner.utils import parallel_state
```
Comment (ZhiyuLi-goog) [inline clm/input_pipeline_gpu.py:5]: Looks good to me given we need additional function call in nemo aligner.

File: clm/model_utils_gpu.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        config,
+        trainer,
+        strict=True,
+        modify_config_fn=_modify_config,
+        restore_path=config.model.restore_from_path,
+        return_updated_cfg=True,
+    )
+    init_peft(model, updated_cfg)
+
+    init_distributed(trainer, model, config.model.get("transformer_engine", False))
```
Comment (ZhiyuLi-goog) [inline clm/model_utils_gpu.py:56]: Look good to me since we need this function to pytorch lightening trainer.

File: clm/model_utils_gpu.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    master_port = os.getenv('MASTER_PORT', '6000')
+    import datetime
+    DEFAULT_TIMEOUT = datetime.timedelta(minutes=60)
+    init_method += master_ip + ':' + master_port
+    torch.distributed.init_process_group(backend='nccl', timeout=DEFAULT_TIMEOUT, world_size=world_size, rank=rank, init_method=init_method)
+    return local_rank, rank, world_size
+
+def setup_model_optimizer(config, trainer):
+    logging.info("loading model")
+    model, updated_cfg = load_from_nemo(
```
Comment (ZhiyuLi-goog) [inline clm/model_utils_gpu.py:45]: Looks good to me if we need `load_from_nemo`.

Alternatively, if we can load from state_dict directly consider the following alternative`:
```
CustomMegatronGPTModel(MegatronGPTModel):
    def load_model_state_dict(self, checkpoint) -> None:
        ...
```
[here](https://github.com/mlcommons/training_results_v3.1/blob/main/NVIDIA/benchmarks/gpt3/implementations/pytorch/custom_callbacks.py#L503-L517)

File: clm/trainer_utils_gpu.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        self.set_max_steps()
+
+        self.run_init_validation = run_init_validation  # do we run validation in the beginning before any training
+        self.logger = logger
+        self.ckpt_callback = ckpt_callback
+        # this timer checks if we should stop training
+        self.run_timer = run_timer
+        
+        # any metrics that require running full token-by-token inference during validation
+        self.inference_metrics_handler = InferenceMetricsHandler(cfg.get("inference_metrics"))
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:91]: Look good to me since we might need this for eval.

File: clm/trainer_utils_gpu.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        self.optimizer.zero_grad()
+
+        self.model.prepare_for_training_step()
+
+        # NOTE: assume backward is called on the loss already
+        loss_mean, metrics = self.model.get_loss_and_metrics(batch=batch, forward_only=False)
+
+        self.model.finish_training_step()
+
+        grad_norm = clip_gradients(self.model, self.cfg.gradient_clip_val)  # .item()
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:151]: LGTM. We need this one `clip_gradients` for distributed optimizer.

File: clm/trainer_utils_gpu.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+        
+        self.args = args
+        self.tokenizer = tokenizer
+        self.optimizer = optimizers[0]
+        self.scheduler = optimizers[1]
+
+        self.step = 0
+        self.consumed_samples = 0
+
+        self.timer = SyncTimer(
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:65]: LGTM.

File: clm/trainer_utils_gpu.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+                metrics["epoch"] = self.epoch + 1
+                self.logger.log_metrics(
+                    metrics, step=self.step, prefix="train/",
+                )
+                metrics = {f"train_{k}": v for k, v in metrics.items()}
+
+                self.step += 1
+
+                run_time_exceeded = self.run_timer.is_finished()
+                run_val, save_model, is_train_end = check_progress(
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:229]: LGTM.

Looks like `check_progress` is used for model saving only. Let's keep it though we probably don't need to save checkpoint in the middle.

File: clm/trainer_utils_gpu.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+        )
+
+    def setup(self, cfg, logger, ckpt_callback, run_timer, run_init_validation=False):
+        self.cfg = cfg
+        # compute `max_steps`
+        self.num_steps_per_epoch = compute_num_steps_per_epoch(
+            self.train_dataloader.batch_sampler, self.cfg.get("limit_train_batches", 1.0)
+        )
+
+        self.limit_val_batches = compute_limit_batches(len(self.val_dataloader), self.cfg.limit_val_batches)
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:76]: LGTM. 

`compute_limit_batches` is used to limit eval step, though we might not need this feature.

File: clm/trainer_utils_gpu.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+        self.consumed_samples = 0
+
+        self.timer = SyncTimer(
+            reduction="mean", sync_cuda=True, buffer_size=1, reduce_op=torch.distributed.ReduceOp.MAX
+        )
+
+    def setup(self, cfg, logger, ckpt_callback, run_timer, run_init_validation=False):
+        self.cfg = cfg
+        # compute `max_steps`
+        self.num_steps_per_epoch = compute_num_steps_per_epoch(
```
Comment (ZhiyuLi-goog) [inline clm/trainer_utils_gpu.py:72]: LGTM.

**Review by ZhiyuLi-goog — COMMENTED** [review]
Thank you @JustinPan-goog! Awesome!

Let's discuss the tradeoff between
* [sft in nemo-aligner](https://github.com/NVIDIA/NeMo-Aligner/blob/d8ef9fb05dc9a30780d5f0092b918476fecd83be/docs/user-guide/sft.rst#L127)
* [pretrain trainer in nemo](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/parts/megatron_trainer_builder.py#L43)

**Comment by ZhiyuLi-goog** [comment]
We might need docker image as the next step including a frozen nemo/nemo-aligner version.

File: clm/gpt_sft_dataset.py (hunk 1 above)
Comment (ZhiyuLi-goog) [reply clm/gpt_sft_dataset.py:1]: The dataset may have its own format:
```
            return {
                'tokens': tokens,
                'labels': labels,
                'attention_mask': self.attention_mask,
                'loss_mask': self.loss_mask,
                'position_ids': self.position_ids,
            }
```
https://github.com/NVIDIA/NeMo/blob/7a3dd6bfe17c0ff16abf04b55c37a5e56f564227/nemo/collections/nlp/data/language_modeling/megatron/gpt_dataset.py#L454-L503
